qcodes/instrument_drivers/yokogawa/GS200.py

```diff
@@ -10,29 +167,343 @@ class GS200(VisaInstrument):
+        # range. Getting the range always returns the cached value. This value is adjusted when calling
+        # self._set_range
+        self._cached_range_value = None
+
+        self.add_parameter('voltage_range',
+                           label='Voltage Source Range',
+                           unit='V',
+                           get_cmd=partial(self._get_range, "VOLT"),
+                           set_cmd=partial(self._set_range, "VOLT"),
+                           vals=Enum(10e-3, 100e-3, 1e0, 10e0, 30e0))
+
+        self.add_parameter('current_range',
+                           label='Current Source Range',
+                           unit='I',
+                           get_cmd=partial(self._get_range, "CURR"),
+                           set_cmd=partial(self._set_range, "CURR"),
+                           vals=Enum(1e-3, 10e-3, 100e-3, 200e-3)
+                           )
+
+        # This is changed through the source_mode interface
+        self.range = self.voltage_range
+
+        self._auto_range = False
+        self.add_parameter('auto_range',
+                           label='Auto Range',
+                           set_cmd=self._set_auto_range,
+                           get_cmd=lambda: self._auto_range,
+                           vals=Bool())
 
         self.add_parameter('voltage',
                            label='Voltage',
                            unit='V',
-                           get_cmd=':SOURce:LEVel?',
-                           set_cmd=':SOURce:LEVel:AUTO {:.4f}',
+                           set_cmd=partial(self._get_set_output, "VOLT"),
+                           get_cmd=partial(self._get_set_output, "VOLT")
+                           )
+
+        self.add_parameter('current',
+                           label='Current',
+                           unit='I',
+                           set_cmd=partial(self._get_set_output, "CURR"),
+                           get_cmd=partial(self._get_set_output, "CURR")
+                           )
+
+        # This is changed through the source_mode interface
+        self.output_level = self.voltage
+
+        self.add_parameter('voltage_limit',
+                           label='Voltage Protection Limit',
+                           unit='V',
+                           vals=Ints(1, 30),
+                           get_cmd=":SOUR:PROT:VOLT?",
+                           set_cmd=":SOUR:PROT:VOLT {}",
+                           get_parser=float_round,
+                           set_parser=int)
+
+        self.add_parameter('current_limit',
+                           label='Current Protection Limit',
+                           unit='I',
+                           vals=Numbers(1e-3, 200e-3),
+                           get_cmd=":SOUR:PROT:CURR?",
+                           set_cmd=":SOUR:PROT:CURR {:.3f}",
                            get_parser=float,
-                           vals=Numbers(-10, 10))
```

Comment:
Looks like the validator disappeared from the voltage was that intentional?

Reply:
Validation is now done in the _set_output function, so this is intentional.